src/app/wallets/reimburse-failed-usd.ts

```diff
@@ -0,0 +1,80 @@
+import { toSats } from "@domain/bitcoin"
+import {
+  CouldNotFindWalletForWalletCurrencyError,
+  InvalidAccountForWalletIdError,
+} from "@domain/errors"
+import { DisplayCurrency, toCents } from "@domain/fiat"
+import { LedgerTransactionType } from "@domain/ledger"
+import { AmountCalculator, WalletCurrency } from "@domain/shared"
+
+import { WalletsRepository } from "@services/mongoose"
+import * as LedgerFacade from "@services/ledger/facade"
+
+const calc = AmountCalculator()
+
+export const reimburseFailedUsdPayment = async <
+  S extends WalletCurrency,
+  R extends WalletCurrency,
+>({
+  paymentFlow,
+  journalId,
+  accountId,
+}: {
+  paymentFlow: PaymentFlow<S, R>
+  journalId: LedgerJournalId
+  accountId: AccountId
+}): Promise<true | ApplicationError> => {
+  const paymentHash = paymentFlow.paymentHashForFlow()
+  if (paymentHash instanceof Error) return paymentHash
+
+  const metadata: FailedPaymentLedgerMetadata = {
+    hash: paymentHash,
+    type: LedgerTransactionType.Payment,
+    pending: false,
+    related_journal: journalId,
+
+    usd: Number(
+      calc.divFloor(paymentFlow.usdPaymentAmount, 100n).amount,
+    ) as DisplayCurrencyBaseAmount,
+
+    satsAmount: toSats(paymentFlow.btcPaymentAmount.amount),
+    centsAmount: toCents(paymentFlow.usdPaymentAmount.amount),
+    satsFee: toSats(paymentFlow.btcProtocolFee.amount),
+    centsFee: toCents(paymentFlow.usdProtocolFee.amount),
+
+    displayAmount: Number(
+      paymentFlow.usdPaymentAmount.amount,
+    ) as DisplayCurrencyBaseAmount,
+    displayFee: Number(paymentFlow.usdProtocolFee.amount) as DisplayCurrencyBaseAmount,
+    displayCurrency: DisplayCurrency.Usd,
+  }
+
+  const txMetadata: LnLedgerTransactionMetadataUpdate = {
+    hash: paymentHash,
+  }
+
+  const { id: senderWalletId } = paymentFlow.senderWalletDescriptor()
+  const recipientWallets = await WalletsRepository().listByAccountId(accountId)
+  if (recipientWallets instanceof Error) return recipientWallets
+  if (!recipientWallets.map((wallet) => wallet.id).includes(senderWalletId)) {
+    return new InvalidAccountForWalletIdError(
+      JSON.stringify({ accountId, walletId: senderWalletId }),
+    )
+  }
+  const btcWallet = recipientWallets.find(
+    (wallet) => wallet.currency === WalletCurrency.Btc,
+  )
+  if (btcWallet === undefined) return new CouldNotFindWalletForWalletCurrencyError()
+  const btcWalletDescriptor = { id: btcWallet.id, currency: btcWallet.currency }
+
+  const result = await LedgerFacade.recordReceive({
+    description: "usd failed payment reimburse",
+    recipientWalletDescriptor: btcWalletDescriptor,
+    amountToCreditReceiver: paymentFlow.totalAmountsForPayment(),
+    metadata,
+    txMetadata,
```

Comment:
it's probably not the point of this PR to change this, but the `metadata` and txMetadata` property is pretty confusing.

Reply:
Yea I think we'd need to clean this up at some point when we sort out how we'll be reading from the `transaction_metadata` table (gql [dataloader](https://github.com/GaloyMoney/galoy/pull/1427#issuecomment-1187422982) planned piece of work).